sdk/go/pulumi/types.go

```diff
@@ -426,7 +429,18 @@ func newApplier(fn interface{}, elemType reflect.Type) (_ *applier, err error) {
 		elemName = "second"
 		fallthrough // validate element type
 	case 1:
-		if t := ft.In(elemIdx); !elemType.AssignableTo(t) {
+		switch t := ft.In(elemIdx); {
+		case elemType.AssignableTo(t):
+			// Do nothing.
+		case elemType.ConvertibleTo(t) && elemType.Kind() == t.Kind():
+			// Support coercion if the types are the same kind.
+			// Types with different internal representations
+			// do not coerce for "free"
+			// (e.g. string([]byte{..}) allocates)
+			// and may not match user expectations
+			// (e.g. string(42) is "*", not "42").
+			ap.convertTo = t
```

Comment:
This comment suggests to me it will make those conversions and could lead to surprising behavior. Should we restrict this to "simple" kinds and report a "nice error" if it's not? e.g.: bool, int(N), uint(N), string

Reply:
Thanks, I will clarify the comment.

RE: simple conversions:
`int` and `uint` conversions sound good, but I'm not sure what you're suggesting for bool/string.
I would suggest against any amount of string parsing/truthiness evaluation (e.g. `int` becomes `bool` via `x > 0`)
because that's too need dependent and users should pick out the behavior they want explicitly.
I will defer these conversions to a separate issue, though. Created #11919 to track.

RE: error message: I recently made a change here to start reporting the invalid applier's location.
We can make the error richer if needed, but I suspect we don't need to augment it just for the 'free' conversions.

Reply:
Ah, I think I misread the comment. I get it - `string` and `[]byte` have different `Kind`, so the conversion **is** prohibited in those cases.